# Python

---

### How to read this notebook

First read the commented CPP cell, that will let you know what you are going to learn in python in the following cell hence giving you some context.

Each topic follows the same two beats:

1. **C++ cell** — the idiom you already know, fully commented out. Do not run it; it is reference, not code.
2. **Python cell** — the runnable equivalent. Run it. It prints its own output, so you can check your prediction against reality.

### How to use google colab
To run a code cell, use shift+Enter key, the cell will execute and you'll see the results below.


> You'll have to solve this in order to get through your first lab. This is an important assignment for your first lab's grades.

---
# 1. Imports, Variables and Control Flow

This section is deliberately fast. You know what a loop is; you need to know
how Python writes one, and which four or five things behave differently from what you expect.

### 1.1 Imports: Python's `#include`

In [2]:
# ---------------- C++ : headers and namespaces (reference only) ----------------
#
#   #include <iostream>      // header brings in declarations
#   #include <vector>
#   #include <cmath>
#
#   using namespace std;     // bring names into scope
#   sqrt(25.0);
#   std::sqrt(25.0);         // or qualify explicitly
#
# A header declares, a .cpp defines, and the linker joins them.
# ------------------------------------------------------------------------------

In [3]:
import math                     # importing whole module, like "#include <cmath>" in CPP
from math import sqrt, pi       # specific names, like `using std::sqrt`
import json as js               # alias, like a namespace alias (using namespace "std")

print(math.sqrt(25))            # qualified
print(sqrt(25), round(pi, 4))   # unqualified, because we imported the names
print(js.dumps({"ok": True}))

# A .py file is both header and source. There is no declaration/definition split

5.0
5.0 3.1416
{"ok": true}


> **Tip.** In Colab, a line starting with `!` is a shell command, not Python:
> `!pip install pydantic` installs a package. Everything in Sections 1 and 2 is built in.
>
> **Why this matters.** You will read `from typing import TypedDict, Annotated` and
> `from pydantic import BaseModel` on nearly every page of framework documentation. Those are
> imports of *names*, not files.

### 1.2 Variables, dynamic typing, truthiness and `None`

In [ ]:
# ---------------- C++ : types are fixed at compile time (reference only) -------
#
#   int age = 25;
#   double gpa = 3.85;
#   bool ok = true;                  // lowercase
#   string name = "Alice";
#
#   age = "twenty-five";             // compile error: the type is part of the variable
#
#   if (v.empty()) { ... }           // you must ask about emptiness explicitly
#   int* p = nullptr;
# ------------------------------------------------------------------------------

In [ ]:
age = 25                  # Variable declaration and assignment
gpa = 3.85                # Floating point variable
ok = True                 # Boolean Variable, True with capital T. `true` is just an undefined name in Python.
name = "Alice"            # String Variable
print(type(age), type(gpa), type(ok), type(name))     # type() function is used to check type of a variable

age = "twenty-five"       # variable age has already been assign but it still is legal: the *name* is re-bound to a new object / previous value overwritten
print(age, type(age))

nothing = None            # Python's nullptr, but a real object you can pass around
print(nothing, type(nothing))

# Truthiness: zero and every empty container are False.
a_list = [0, 1, "", "x", [], [0], {}, None]    # creating a list of items, we'd loop on this to check truthiness or falsiness of each to know what eachone is
for value in a_list:
    print(f"  bool({value}) = {bool(value)}")

# try changing this -> add 0.0, " ", or (0,) to the tuple above and predict each result

<class 'int'> <class 'float'> <class 'bool'> <class 'str'>
twenty-five <class 'str'>
None <class 'NoneType'>
  bool(0) = False
  bool(1) = True
  bool() = False
  bool(x) = True
  bool([]) = False
  bool([0]) = True
  bool({}) = False
  bool(None) = False


> **Trap.** `True` and `False` are capitalised. Writing `if ok == true:` raises `NameError`.
>
> **Trap.** Empty containers are falsy, so `if a_list:` is the idiomatic way to say "not empty",
> replacing `if (!v.empty())`. But note `bool([0])` is `True`: the list has one element, and the
> value of that element is irrelevant.
>
> **Tip.** Compare against `None` with `is None`, never `== None`. `is` compares identity,
> like comparing pointers.

### 1.3 Strings and f-strings

In [ ]:
# ---------------- C++ : std::string (reference only) --------------------------
#
#   #include <string>
#   std::string s = "Hello, World!";
#
#   s.length();               // 13
#   s.substr(0, 5);           // "Hello"
#   s.find("World");          // 7, or npos
#   s[0] = 'h';               // legal: std::string is mutable
#
# ------------------------------------------------------------------------------

In [ ]:
s = "Hello, World!"

print(len(s), s[0], s[-1])          # -1 indexes from the end of the collection
print(s[0:5])                       # To get a substring, starts at 0th index ends at 4th. stop index is EXCLUSIVE, hence giving output 'Hello'
print(s[7:])                        # If you don't give ending index, that becomes end of the collection, including last character
print(s[::-1])                      # Python trick to reverese a string
print(s.upper(), s.split(", "), s.replace("o", "0"))   # These methods have pretty self explanatory names
print(repr("  padded  ".strip()))

name, age = "Alice", 25
print(f"Name: {name}, Age: {age}")  # f-string, let you interpolate variable values on the go

# s[0] = "h"                        # TypeError: strings are immutable

13 H !
Hello
World!
!dlroW ,olleH
HELLO, WORLD! ['Hello', 'World!'] Hell0, W0rld!
'padded'
Name: Alice, Age: 25


> **Trap.** Python strings are immutable. `s[0] = "h"` raises `TypeError`, unlike C++ where it
> is fine. Every "modifying" method returns a *new* string, so `s.upper()` alone changes nothing;
> you must assign the result.
>
> **Trap.** Slice stop indices are exclusive: `s[0:5]` gives indices 0 through 4.
>
> **Why this matters.** f-strings are how you will assemble prompts. Get comfortable with them now.

### 1.4 Conditionals and indentation

In [ ]:
# ---------------- C++ : braces delimit blocks (reference only) ----------------
#
#   if (score >= 90) {
#       grade = 'A';
#   } else if (score >= 80) {
#       grade = 'B';
#   } else {
#       grade = 'F';
#   }
#
#   std::string status = (score >= 60) ? "pass" : "fail";
#   if (x > 10 && x < 20) { ... }        // && || !
# ------------------------------------------------------------------------------

In [ ]:
# Python uses indentation to define a code block and not braces, if not followed, will give error
score = 85

if score >= 90:
    grade = "A"
elif score >= 80:          # elif, not "else if"
    grade = "B"
else:
    grade = "F"
print(grade)

# conditional expression while defining a variable unlike cpp ( though ternary operator does that in cpp )
status = "pass" if score >= 60 else "fail"
print(status)

# Logical Operators (and or not) words, not symbols (&& || !)
x = 15
print(x > 10 and x < 20)
print(14 < x < 20)         # chained comparison, no C++ equivalent, x is being compared with two other values

B
pass
True
True


> **Trap.** Indentation *is* the block delimiter. There are no braces, and misaligned code is an
> `IndentationError` rather than a silently different program. Use 4 spaces and never mix tabs
> with spaces.
>
> **Trap.** Logical operators are `and`, `or`, `not`. Writing `&&` is a `SyntaxError`; `&` and `|`
> exist but mean bitwise-and and bitwise-or, so `if a & b:` may run and give a wrong answer.

### 1.5 Loops: `for` is always a for-each

In [ ]:
# ---------------- C++ : three loop forms (reference only) --------------------
#
#   for (int i = 0; i < 5; i++) { ... }            // C-style, index-driven
#
#   vector<string> fruits = {"apple", "banana"};
#   for (auto f : fruits) { ... }           // range-based, C++11
#
#   int n = 0;
#   while (n < 5) { cout << n++; }
# ------------------------------------------------------------------------------

In [ ]:
for i in range(5):                 # 0,1,2,3,4  -- range(n) is [0, n) n being exclusive
    print(i, end=" ")             # end=" " is specific delimiter used to define that don't go new line after printing an i and just end with a space, because by default, it goes on new line.
print()

for i in range(2, 11, 2):          # start, stop, step(jump) -> 2,4,6,8,10
    print(i, end=" ")
print()

fruits = ["apple", "banana", "cherry"]
for f in fruits:                   # same as the C++11 range-for
    print(f, end=" ")
print()

n = 0
while n < 3:
    print(n, end=" ")
    n += 1                         # there is no n++ in Python
print()

# Have a look at zip and enumerate functions, what they are and how they operate.

0 1 2 3 4 
2 4 6 8 10 
apple banana cherry 
0 1 2 


> **Trap.** Python has no `++` or `--`. Use `n += 1` and `n -= 1`.
>
> **Trap.** There is no C-style `for`. Index-based loops go through `range()`, and if you find
> yourself writing `for i in range(len(items)):` just to read `items[i]`, use `enumerate()` instead.
>
> **Tip.** `range(1, n + 1)` gives you the inclusive `[1, n]` that C++ programmers usually want.

### 1.6 Comprehensions (Optional but so powerful)

In [ ]:
# ---------------- C++ : build then fill (reference only) ---------------------
#
#   vector<int> squares;
#   for (int i = 1; i <= 5; i++)
#       squares.push_back(i * i);
#
#   // C++20 ranges::views can do this in one expression, but verbosely.
# ------------------------------------------------------------------------------

In [ ]:
# [ expression for variable in iterable if condition ]
squares = [x**2 for x in range(1, 6)]    # they are a little hard to comprehend for first time, so give them due time.
print(squares)

evens = [x for x in range(10) if x % 2 == 0]
print(evens)

matrix = [[1, 2], [3, 4], [5, 6]]
flat = [v for row in matrix for v in row]      # nested loops, outer first
print(flat)

[1, 4, 9, 16, 25]
[0, 2, 4, 6, 8]
[1, 2, 3, 4, 5, 6]


> **Tip.** A comprehension is the normal way to transform a sequence in Python.

1.   List item
2.   List item

Reach for it
> before `map` or `filter`; it reads better and is usually faster.
>
> **Trap.** Comprehensions are for building values. If the loop body is doing work rather than
> producing an element, write an ordinary `for` loop.

### 1.7 Functions: defaults, keyword arguments, `*args` and `**kwargs`

In [ ]:
# ---------------- C++ : typed signatures and overloading (reference only) ----
#
#   string greet(string name) { return "Hello, " + name; }
#
#   int power(int base, int exp = 2) { return (int)pow(base, exp); }
#   power(3);        // 9
#   power(3,2);     // 9
#
#   int    add(int a, int b)       { return a + b; }     // overload set:
#   double add(double a, double b) { return a + b; }     // same name, many signatures
# ------------------------------------------------------------------------------

In [1]:
def greet(name):
    return f"Hello, {name}!"

def power(base, exp=2):            # default argument, same idea as C++
    return base ** exp             # ** is exponentiation, no pow()

print(greet("Alice"))
print(power(3), power(3, 3))

def add_all(*nums):                # variadic positional; nums is a tuple
    return sum(nums)

def describe(**info):              # variadic keyword; info is a dict
    for key, value in info.items():
        print(f"  {key} = {value}")

print(add_all(1, 2, 3, 4))
describe(name="Alice", age=25, active=True)

Hello, Alice!
9 27
10
  name = Alice
  age = 25
  active = True


> **Trap.** Python has no function overloading. A second `def add(...)` simply replaces the first.
> Use default arguments, `*args` / `**kwargs`, or an `isinstance()` check instead.
>
> **Why this matters.** Framework calls are keyword-heavy:
> `llm.invoke(prompt, temperature=0.7, max_tokens=512)`. Reading a signature full of
> `*args, **kwargs` is a daily activity, so make sure you can tell the tuple from the dict.

---
### Check yourself: Section 1

**1.** What does `bool([0])` return?

<details><summary>1</summary>

`True`. The list contains one element, so it is not empty. Truthiness looks at the container's
length, not at the values inside it.
</details>

**2.** `s = "abc"` followed by `s[0] = "A"` — what happens?

<details><summary>Answer</summary>

`TypeError`. Strings are immutable. You would write `s = "A" + s[1:]` instead.
</details>

**3.** What does `range(1, 5)` produce, and how many iterations does `for i in range(1, 5)` run?

<details><summary>Answer</summary>

`1, 2, 3, 4` — four iterations. The stop value is exclusive.
</details>

**4.** Given `def power(base, exp=2)`, what do `power(3, 3)` and `power(exp=3, base=2)` return?

<details><summary>Answer</summary>

`27` and `8`. In the second call the keywords bind the arguments by name, so the positional
order is irrelevant.
</details>

### Question 1: FizzBuzz (Pythonic Style)
Write a program that prints the numbers from 1 to 100. For multiples of three, print "Fizz" instead of the number, and for the multiples of five, print "Buzz". For numbers which are multiples of both three and five, print "FizzBuzz".

**Challenge:** solve it once using list comprehension and once through a single `if/elif/else` block.

In [18]:
# write your code here (if/elif/else)
for i in range(1, 101):
    if i % 3 == 0 and i % 5 == 0:
        print("FizzBuzz", end=" ")
    elif i % 3 == 0:
        print("Fizz", end=" ")
    elif i % 5 == 0:
        print("Buzz", end=" ")
    else:
        print(i, end=" ")

# comprehension
result = ["FizzBuzz" if i % 15 == 0 else "Fizz" if i % 3 == 0 else "Buzz" if i % 5 == 0 else str(i)
         for i in range(1,101)]
print(" ".join(result))

1 2 Fizz 4 Buzz Fizz 7 8 Fizz Buzz 11 Fizz 13 14 FizzBuzz 16 17 Fizz 19 Buzz Fizz 22 23 Fizz Buzz 26 Fizz 28 29 FizzBuzz 31 32 Fizz 34 Buzz Fizz 37 38 Fizz Buzz 41 Fizz 43 44 FizzBuzz 46 47 Fizz 49 Buzz Fizz 52 53 Fizz Buzz 56 Fizz 58 59 FizzBuzz 61 62 Fizz 64 Buzz Fizz 67 68 Fizz Buzz 71 Fizz 73 74 FizzBuzz 76 77 Fizz 79 Buzz Fizz 82 83 Fizz Buzz 86 Fizz 88 89 FizzBuzz 91 92 Fizz 94 Buzz Fizz 97 98 Fizz Buzz 1 2 Fizz 4 Buzz Fizz 7 8 Fizz Buzz 11 Fizz 13 14 FizzBuzz 16 17 Fizz 19 Buzz Fizz 22 23 Fizz Buzz 26 Fizz 28 29 FizzBuzz 31 32 Fizz 34 Buzz Fizz 37 38 Fizz Buzz 41 Fizz 43 44 FizzBuzz 46 47 Fizz 49 Buzz Fizz 52 53 Fizz Buzz 56 Fizz 58 59 FizzBuzz 61 62 Fizz 64 Buzz Fizz 67 68 Fizz Buzz 71 Fizz 73 74 FizzBuzz 76 77 Fizz 79 Buzz Fizz 82 83 Fizz Buzz 86 Fizz 88 89 FizzBuzz 91 92 Fizz 94 Buzz Fizz 97 98 Fizz Buzz


---
# 2. Data Structures and the Aliasing Trap

The container mapping is quick, and you can skim it. Then comes 2.4, which is
the single largest semantic difference between C++ and Python and the one most likely to produce
a bug you cannot explain.

### 2.1 `list` is `std::vector`

In [ ]:
# ---------------- C++ : std::vector (reference only) -------------------------
#
#   std::vector<int> v = {3, 1, 4, 1, 5};
#   v[0];  v.back();
#   v.push_back(9);
#   v.insert(v.begin(), 0);
#   v.erase(v.begin() + 2);
#   v.size();
#   std::sort(v.begin(), v.end());
#
#   vector<int> sub(v.begin() + 1, v.begin() + 4);   // no slice syntax
# ------------------------------------------------------------------------------

In [ ]:
v = [3, 1, 4, 1, 5]

print(v[0])
print(v[-1])          # -1 replaces v.back()
print(v[1:4])      # slice
print(v[::-1])     # and reversed copy
v.append(9)                 # push_back
v.insert(0, 0)              # insert at index
v.remove(1)                 # remove FIRST match BY VALUE, not by index
last = v.pop()              # remove and return the last element
print(v, " popped:", last, " len:", len(v))

v.sort()                    # in place
print(v)
print(sorted(v, reverse=True))    # sorted() returns a new list
print([1, "two", 3.0, True])          # heterogeneous by default, supports multiple data types in one list

3
5
[1, 4, 1]
[5, 1, 4, 1, 3]
[0, 3, 4, 1, 5]  popped: 9  len: 5
[0, 1, 3, 4, 5]
[5, 4, 3, 1, 0]
[1, 'two', 3.0, True]


> **Trap.** `v.remove(x)` removes by *value*; `del v[i]` and `v.pop(i)` remove by *index*.
> Mixing them up on a list of integers gives silently wrong results.


### 2.2 `dict` is `std::unordered_map`

In [ ]:
# ---------------- C++ : std::unordered_map (reference only) ------------------
#
#   unordered_map<string, int> ages;
#   ages["Alice"] = 25;
#   ages["Alice"];                 // operator[] default-constructs if absent
#   ages.at("Carol");              // throws if absent
#   ages.count("Carol");           // 0 or 1
#   ages.erase("Bob");
#
#   for (const auto& [k, v] : ages) { ... }
# ------------------------------------------------------------------------------

In [20]:
ages = {"Alice": 25, "Bob": 30}

print(ages["Alice"])
print(ages.get("Carol", 0))      # safe read with a default; no KeyError
print("Bob" in ages)             # membership tests the KEY

ages["Carol"] = 28               # insert or overwrite
del ages["Bob"]
print(ages)

for key, value in ages.items():
    print(f"  {key}: {value}")


25
0
True
{'Alice': 25, 'Carol': 28}
  Alice: 25
  Carol: 28


> **Trap.** `d["missing"]` raises `KeyError` — it behaves like `.at()`, not like C++'s
> `operator[]`, and never inserts a default. Use `d.get(key, default)` whenever the key might
> be absent.
>
> **Tip.** Since Python 3.7, `dict` preserves insertion order, so it covers both
> `unordered_map` and the common uses of `std::map`.
>
> **Why this matters.** Framework configuration, model responses, and tool arguments are all
> dicts. `.get()` with a default is the difference between a graceful fallback and a crash.

### 2.3 `tuple` and `set`

In [ ]:
# ---------------- C++ : std::pair, std::tuple, std::unordered_set (ref only) --
#
#   std::pair<int,int> coords = {10, 20};
#   auto [x, y] = coords;                       // C++17 structured binding
#
#   std::unordered_set<int> s = {1, 2, 3};
#   s.insert(4);  s.erase(2);  s.count(3);
#   // set_union / set_intersection need <algorithm> and iterators
# ------------------------------------------------------------------------------

In [24]:
coords = (10, 20)          # tuple: immutable sequence
x, y = coords              # unpacking, like a structured binding
print(x, y)

def min_max(values):
    return min(values), max(values)      # returning two things builds a tuple

lo, hi = min_max([3, 1, 4, 1, 5])
print(lo, hi)

single = (42,)             # a ONE-element tuple needs the trailing comma
print(type(single), type((42)))          # (42) is just the integer 42

s = {1, 2, 3}              # set literal
s.add(4)
s.discard(99)              # silent if absent; s.remove(99) would raise
print(s, 3 in s)

a, b = {1, 2, 3, 4}, {3, 4, 5}
print(a | b)               # union, intersection, difference
print(a & b)
print(a - b)

print(sorted(set([1, 1, 2, 3, 3])))      # deduplicate a list, remove duplicates

10 20
1 5
<class 'tuple'> <class 'int'>
{1, 2, 3, 4} True
{1, 2, 3, 4, 5}
{3, 4}
{1, 2}
[1, 2, 3]


> **Trap.** `{}` is an empty **dict**, not an empty set. Use `set()` for that.
>
> **Trap.** `(42)` is an integer; the parentheses are just grouping. Only the comma makes a tuple.

### 2.4 Assignment is aliasing, not copying

In [ ]:
# ---------------- C++ : copies by default (reference only) -------------------
#
#   std::vector<int> a = {1, 2, 3};
#   std::vector<int> b = a;      // COPY: the copy constructor runs
#   b.push_back(4);              // a is unchanged, still {1, 2, 3}
#
#   std::vector<int>& r = a;     // you must ASK for a reference
#   r.push_back(4);              // now a changes
# ------------------------------------------------------------------------------

In [26]:
a = [1, 2, 3]
b = a                  # NOT a copy: two names for ONE list object
b.append(4)            # Appended to a too, why? they are not two different objects, b = a makes b point where a is pointing, Not a copy.
#d = [1, 2, 3]
print("a =", a)                    # [1, 2, 3, 4] -- a changed too
print("b =", b)
print("same object?", a is b)      # True
print(id(a) == id(b))              # identical addresses

c = a.copy()           # explicit copy. list(a) and a[:] also work.
c.append(99)
print("after copying: a =", a, " c =", c)

a = [1, 2, 3, 4]
b = [1, 2, 3, 4]
same object? True
True
after copying: a = [1, 2, 3, 4]  c = [1, 2, 3, 4, 99]


> **Trap.** This is the most important cell in the notebook. In C++ `b = a` copies; in Python it
> binds a second name to the same object. Every Python variable behaves like a pointer that is
> dereferenced automatically, and there is no copy constructor being called on your behalf.
>
> Two consequences:
>
> - Passing a list or dict into a function and mutating it changes the caller's object. There is
>   no `const&` to prevent this.
> - This never bites with `int`, `str`, `float`, or `tuple`, because those are immutable — there
>   is no way to mutate them, so sharing is invisible.
>
> **Tip.** When you need a real copy: `list(x)`, `dict(x)`, `x.copy()` for one level, and
> `copy.deepcopy(x)` when the structure is nested.

### 2.5 Unpacking with `*` and `**`, and nested data

In [27]:
def connect(host, port, timeout=10):
    print(f"  {host}:{port} timeout={timeout}")

args = ["localhost", 8080]
config = {"host": "example.com", "port": 443, "timeout": 30}

connect(*args)          # spread a sequence into positional parameters
connect(**config)       # spread a dict into keyword parameters

defaults = {"temperature": 0.8, "max_tokens": 256}
overrides = {"temperature": 0.7}
merged = {**defaults, **overrides}      # later keys win
print(merged)

# A model response is nested dicts and lists. Index when the key must exist,
# and use .get() chains when it might not.
response = {"choices": [{"message": {"content": "Hello there"}}]}
print(response["choices"][0]["message"]["content"])
print(response.get("usage", {}).get("total_tokens", 0))     # 0, no crash

  localhost:8080 timeout=10
  example.com:443 timeout=30
{'temperature': 0.7, 'max_tokens': 256}
Hello there
0


> **Why this matters.** `**kwargs` pass-through is how options travel down through framework
> layers: a wrapper accepts `**kwargs` and forwards them untouched to the model client. And
> `.get()` chains are what keep your code alive when a provider omits a field you expected.

---
### Check yourself: Section 2

**1.** After `a = [1, 2]`, `b = a`, `b.append(3)` — what is `a`?

<details><summary>Answer</summary>

`[1, 2, 3]`. `b = a` created a second name for the same list, not a copy. `a is b` is `True`.
</details>

**2.** Why does `d = {}` not give you an empty set, and what does?

<details><summary>Answer</summary>

`{}` is the empty dict literal; the braces were taken by dicts first. Use `set()`.
</details>


**3.** What does `{**{"a": 1}, **{"a": 2}}` evaluate to?

<details><summary>Answer</summary>

`{"a": 2}`. When keys collide during unpacking, the rightmost value wins — which is exactly why
this idiom works for overriding defaults.
</details>

### Question 2: Frequency Counter
Take a long string of text and return a `dict` where keys are words and values are how many times they appear.

**Hint:** Use `d.get(word, 0) + 1` to handle missing keys efficiently.

In [8]:
# write your code here
def count_words(text):
    count = {}
    for i in text.split():
        count[i] = count.get(i,0) + 1
    return count

text = "My name is Faizan Asf and what is your name bro/sis"
word_count = count_words(text)
word_count

{'My': 1,
 'name': 2,
 'is': 2,
 'Faizan': 1,
 'Asf': 1,
 'and': 1,
 'what': 1,
 'your': 1,
 'bro/sis': 1}

---
# 3. Classes and Type Hints as Machinery

You will write very few classes when building agents: graph nodes are plain
functions and tools are decorated functions. What you will write constantly is a class body that
*describes the shape of some data* so that a framework can read it.

So this section spends about seven minutes on class syntax — just enough that the type-hint code
below is readable — and the rest on the part that actually matters.

**The central idea.** In C++, types exist for the compiler and are gone at runtime. In Python
they are *runtime data*, and agent frameworks read them: to validate arguments, to merge state
between graph nodes, and to generate the JSON schema that tells a language model how to call
your function.

### 3.1 Classes: `__init__`, `self`, `__repr__`

In [ ]:
# ---------------- C++ : class with constructor (reference only) --------------
#
#   class Dog {
#   public:
#       static std::string species;              // class variable
#       std::string name;
#       int age;
#
#       Dog(const std::string& n, int a) : name(n), age(a) {}   // constructor
#
#       std::string bark() const {               // `this` is implicit
#           return name + " says Woof";
#       }
#   };
#   Dog* d = new Dog("Rex", 5);                  // or Dog d("Rex", 5);
# ------------------------------------------------------------------------------

In [ ]:
class Dog:
    species = "Canis lupus familiaris"      # class variable, shared by all instances

    def __init__(self, name, age):          # the constructor
        self.name = name                    # `self` is C++'s `this`...
        self.age = age                      # ...but it is never implicit, you have to pass it in constructor

    def bark(self):                         # every instance method takes self first
        return f"{self.name} says Woof"

    def __repr__(self):                     # what you see when debugging
        return f"Dog(name={self.name!r}, age={self.age})"

d = Dog("Rex", 5)              # no `new` keyword, and nothing to delete
print(d.bark())
print(d)                       # uses __repr__
print(Dog.species)
print(d.__dict__)              # attributes are just a dict under the hood

Rex says Woof
Dog(name='Rex', age=5)
Canis lupus familiaris
{'name': 'Rex', 'age': 5}


> **Trap.** `self` must be written as the first parameter of every instance method. Omitting it
> gives you `TypeError: bark() takes 0 positional arguments but 1 was given`, which is one of the
> most common beginner errors.
>
> **Trap.** There is no `new`, no `delete`, and no destructor you should rely on. Objects are
> garbage collected, so there is no RAII. Section 6 covers the replacement.
>
> **Tip.** Always write `__repr__` on classes you will debug. Without it, printing an object
> gives you `<__main__.Dog object at 0x7f...>`, which tells you nothing.
>
> **Tip.** There is no enforced access control. A leading underscore, `self._balance`, is a
> convention meaning "internal, do not touch"; nothing stops you.

### 3.2 Type hints



In [ ]:
# ---------------- C++ : the compiler enforces types (reference only) ---------
#
#   int add(int a, int b) { return a + b; }
#   add("con", "cat");     // compile error, caught before the program runs
# ------------------------------------------------------------------------------

In [ ]:
from typing import Any

def add(a: int, b: int) -> int:
    return a + b

print(add(2, 3))
print(add("con", "cat"))        # NOT an error: hints are not enforced at runtime

def process(items: list[str], config: dict[str, Any]) -> int:
    return len(items)

print(process(["a", "b"], {"mode": "fast"}))

print(add.__annotations__)      # the hints are readable data, attached to the function

5
concat
2
{'a': <class 'int'>, 'b': <class 'int'>, 'return': <class 'int'>}


> **Trap.** A type hint is documentation plus metadata. Nothing checks it while your program
> runs, so `add("con", "cat")` happily returns `"concat"`. Static checkers such as `mypy` will
> flag it, but they are a separate tool you have to run.
>
> **Why this matters.** Notice that last line. `add.__annotations__` is an ordinary dict that
> any library can read. That is the whole trick behind everything in the rest of this section:
> your hints are data, and the framework inspects them.

### 3.3 Pydantic: hints that do get enforced

> Pydantic is Python's standard library for runtime data validation, parsing, and settings management using standard type annotations.


>Runtime Type Enforcement: Unlike native Python type hints—which are strictly static—Pydantic actively validates data during execution and raises a detailed ValidationError if the data doesn't match the schema.

In [ ]:
!pip install -q pydantic

In [34]:
from pydantic import BaseModel, Field, ValidationError

class SearchQuery(BaseModel):
    query: str = Field(description="The text to search for")
    max_results: int = Field(default=5, ge=1, le=50)     # ge/le are constraints

q = SearchQuery(query="python tutorials")
print(q)                            # __repr__ comes for free
print(q.query, q.max_results)       # dotted access, unlike TypedDict

q2 = SearchQuery(query="x", max_results="12")            # string is COERCED to int
print(q2.max_results, type(q2.max_results))

try:
    SearchQuery(query="x", max_results=999)              # violates le=50
except ValidationError as e:
    print("rejected:", e.errors())

schema = SearchQuery.model_json_schema()
print(schema["properties"]["query"]["description"])

query='python tutorials' max_results=5
python tutorials 5
12 <class 'int'>
rejected: [{'type': 'less_than_equal', 'loc': ('max_results',), 'msg': 'Input should be less than or equal to 50', 'input': 999, 'ctx': {'le': 50}, 'url': 'https://errors.pydantic.dev/2.8/v/less_than_equal'}]
The text to search for


> **Tip.** Pydantic is the one place where annotations are checked. Construct a model with the
> wrong type and you get a `ValidationError` at the boundary, which is as close to compile-time
> safety as this ecosystem offers.
>
> **Why this matters.** The last two lines are the point. That `description` string is not a
> comment for humans; it is exported into a JSON schema and sent to the language model so it
> knows what the argument means. A vague description produces a model that calls your tool
> incorrectly. When you write agent tools, the prose in `Field(description=...)` is part of the
> prompt.
>
> **Tip.** `TypedDict` for graph state, Pydantic for anything crossing a boundary — tool
> arguments, structured output, API payloads. `@dataclass` in the standard library is a third,
> lighter option you will meet in other people's code; it validates nothing.

---
### Check yourself: Section 3

**1.** Why does `add("con", "cat")` succeed when `add` is annotated `(a: int, b: int) -> int`?

<details><summary>Answer</summary>

Because annotations are never checked at runtime. They are stored in `add.__annotations__` as
data. Only a separate static checker, or a library like Pydantic, acts on them.
</details>

**2.** In `Field(description="The text to search for")`, who reads that string?

<details><summary>Answer</summary>

The language model. It is exported into the tool's JSON schema and sent along with the request,
so it functions as part of the prompt rather than as a code comment.
</details>

### Question 3: Data Validator (Pydantic)
Create a Pydantic model for a `User` that requires:
*   `username` (str)
*   `email` (str)
*   `age` (int between 0 and 120)
*   `is_active` (bool, default True)

Try to initialize it with valid and invalid data to see the validation in action.

In [1]:
# write your code here
from pydantic import BaseModel, Field, ValidationError
class User(BaseModel):
    username: str
    email: str
    age: int = Field(ge=0,le=120)
    is_active: bool = True

user1 = User(
    username="feezi",
    email="kjhak@gmail.com",
    age=23,
    is_active=True
    ) 
print(user1)

try:
    User(
    username="feezi",
    email="kjhak@gmail.com",
    age="twenty three",
    is_active=True
    )
except ValidationError as e:
    print("Bhaiyaaa Dehaan sa",e.errors())
try:
    User(
    username="feezi",
    email="kjhak@gmail.com",
    age=2,
    is_active=True
    )
except ValidationError as e:
    print("Bhaiyaaa Dehaan sa",e.errors())


username='feezi' email='kjhak@gmail.com' age=23 is_active=True
Bhaiyaaa Dehaan sa [{'type': 'int_parsing', 'loc': ('age',), 'msg': 'Input should be a valid integer, unable to parse string as an integer', 'input': 'twenty three', 'url': 'https://errors.pydantic.dev/2.8/v/int_parsing'}]


---
# 4. Functions as Objects and Decorators

The goal here is narrow: by the end you should be able to look at `@tool`
above a function and know exactly what it does, rather than treating it as syntax you have to
memorise.

### 4.1 Functions are values

In [ ]:
# ---------------- C++ : function pointers and lambdas (reference only) ------
#
#   auto square = [](int x) -> int { return x * x; };
#
#   int offset = 10;
#   auto add_offset = [&offset](int x) { return x + offset; };   // capture list
#
#   std::sort(v.begin(), v.end(),
#             [](const auto& a, const auto& b){ return a.second < b.second; });
# ------------------------------------------------------------------------------

In [53]:
def square(x):
    return x * x

f = square                      # no & needed; the name already refers to the object
print(f(5), f.__name__)

def apply_twice(fn, x):         # functions passed as arguments
    return fn(fn(x))
print(apply_twice(square, 3))

nums = [1, 2, 3, 4, 5]
print([x * 2 for x in nums])                            # preferred
print(list(map(lambda x: x * 2, nums)))                 # std::transform
print(list(filter(lambda x: x % 2 == 0, nums)))         # std::copy_if

students = [("Alice", 90), ("Bob", 75), ("Cara", 82)]
students.sort(key=lambda s: s[1])       # key= is the sort idiom
print(students)

25 square
81
[2, 4, 6, 8, 10]
[2, 4, 6, 8, 10]
[2, 4]
[('Bob', 75), ('Cara', 82), ('Alice', 90)]


> **Tip.** A `lambda` in Python holds a single expression, with no braces and no `return`. If you
> need more than one line, define a named function; that is normal and idiomatic, not a
> workaround.

### 4.3 Writing a decorator

In [55]:
import functools

def logged(fn):                              # takes a function...
    @functools.wraps(fn)                     # ...keeps the original name and docstring
    def wrapper(*args, **kwargs):            # ...builds a replacement...
        print(f"  calling {fn.__name__}{args}")
        result = fn(*args, **kwargs)
        print(f"  -> {result}")
        return result
    return wrapper                           # ...and returns it

@logged
def multiply(a, b):
    return a * b

multiply(3, 4)
print(multiply.__name__)         # 'multiply', thanks to functools.wraps


  calling multiply(3, 4)
  -> 12
multiply


> **Tip.** That comment at the bottom is the whole concept. `@something` above a `def` means
> "pass this function to `something` and rebind the name to whatever comes back". Nothing more.
>
> **Tip.** Always apply `functools.wraps`. Without it the wrapper's own name and docstring
> replace the original's, which breaks the schema generation you are about to see.

### 4.4 Reading `@tool`

In [ ]:
def tool(fn):
    """
    A toy version of the real @tool decorator, to show what it reads.
    It does not wrap the function at all -- it just attaches metadata.
    """
    fn.is_tool = True
    fn.tool_name = fn.__name__
    fn.tool_description = fn.__doc__          # the DOCSTRING becomes the description
    fn.tool_args = fn.__annotations__         # the HINTS become the argument schema
    return fn

@tool
def get_weather(city: str) -> str:
    """Return the current weather for a given city."""
    return f"It is sunny in {city}."

print(get_weather("Lahore"))                  # still an ordinary callable function
print(get_weather.is_tool, "|", get_weather.tool_name)
print(get_weather.tool_description)
print(get_weather.tool_args)

It is sunny in Lahore.
True | get_weather
Return the current weather for a given city.
{'city': <class 'str'>, 'return': <class 'str'>}


> **Why this matters.** The real `@tool` in LangChain does more work, but it reads the same two
> things: your type hints and your docstring. It turns them into a JSON schema and sends that to
> the model, which is how the model knows your function exists, what it is for, and what to pass.
>
> The practical consequence is that in agent code a docstring is not a comment. It is
> instructions to the model, and a sloppy one produces a tool the model misuses.

### Question 4: Timer Decorator
Write a decorator called `@time_it` that calculates and prints how long a function took to execute using `time.perf_counter()`. Apply it to a function that calculates a large factorial or sleeps for a second.

In [66]:
# write your code here
import functools
import time

def time_it(fn):
    @functools.wraps(fn)
    def wrapper(*args, **kwargs):
        start = time.perf_counter()
        result = fn(*args, **kwargs)
        end = time.perf_counter()
        print(f"{fn.__name__} Function has taken {end - start} seconds")
        return result
    return wrapper
@time_it
def slow_function():
    time.sleep(1)

@time_it
def factorial(n):
    result = 1
    for i in range(1,n+1):
        result *= i
    return result
slow_function()
print(factorial(3))

slow_function Function has taken 1.0001360000023851 seconds
factorial Function has taken 8.099999831756577e-06 seconds
6


---
# 5. async / await
Along with Section 3, this is the part of the notebook that will pay off
immediately. Modern framework APIs come in pairs — `invoke` and `ainvoke`, `stream` and
`astream` — and the `a` versions are the ones used in any real application.

**Why frameworks are async.** An agent spends almost all its time waiting on the network: a
model call, a database query, a web search. This is I/O-bound work, not CPU-bound work. In C++
you would reach for `std::thread` or `std::async` and pay for real OS threads, locks, and shared
state. Python's `asyncio` instead runs a single thread with an event loop: when a coroutine hits
a network wait, it hands control back to the loop, which runs something else until the response
arrives. One thread, no locks, and hundreds of calls in flight.

The rule of thumb: `asyncio` for waiting, threads for computing.

### 5.1 `async def`, coroutines, and the missing `await`

In [ ]:
# ---------------- C++ : futures and threads (reference only) ----------------
#
#   #include <future>
#   auto fut = std::async(std::launch::async, fetch, "a");   // may spawn a thread
#   std::string result = fut.get();                          // blocks until ready
#
#   // C++20 coroutines exist but need library support to be usable.
# ------------------------------------------------------------------------------

In [ ]:
import asyncio
import time

async def fetch(name, delay):            # `async def` makes this a coroutine function
    print(f"  start {name}")
    await asyncio.sleep(delay)           # yields control to the loop; does NOT block
    print(f"  done  {name}")
    return f"{name}-result"

coro = fetch("a", 0.1)                   # calling it runs NOTHING yet
print(type(coro))                        # a coroutine object, not a result

result = await coro                      # await runs it and unwraps the value
print(result)

<class 'coroutine'>
  start a
  done  a
a-result


> **Trap.** Calling an `async` function does not run it. It builds a coroutine object. Forget the
> `await` and you get that object where you expected a string, and Python only warns you with
> `RuntimeWarning: coroutine ... was never awaited`. If a value prints as
> `<coroutine object ...>`, you have found your bug.
>
> **Trap.** `await` is only legal inside an `async def` — except in a notebook, where the cell is
> already running inside a loop, which is why the cell above works. In a `.py` script you enter
> async code through `asyncio.run(main())`. Calling `asyncio.run()` inside Colab raises
> `RuntimeError: asyncio.run() cannot be called from a running event loop`, so use bare `await`
> here and `asyncio.run()` in your scripts.

---
### Check yourself: Section 5

**1.** `r = fetch("a", 1)` where `fetch` is an `async def`. What is in `r`?

<details><summary>Answer</summary>

A coroutine object. Nothing has executed. You need `r = await fetch("a", 1)` to get the value.
</details>

**2.** Why does `asyncio.gather` help with two model calls but not with two heavy matrix multiplications?

<details><summary>Answer</summary>

`asyncio` interleaves work on one thread, so it only wins when tasks spend their time *waiting*.
Model calls wait on the network. Matrix multiplication needs CPU, and a single thread cannot do
two of those at once — that needs real threads or processes.
</details>

**3.** Why does `asyncio.run(main())` fail in Colab but work in a `.py` file?

<details><summary>Answer</summary>

The notebook is already running inside an event loop, and you cannot start a second one in the
same thread. In a script there is no loop yet, so `asyncio.run` creates one. In the notebook,
`await` directly.
</details>

### Question 5: Mock API Fetcher
Write an `async` function that simulates fetching data from three different URLs (using `asyncio.sleep` to simulate lag).

In [3]:
import asyncio

async def fetch_url(url, delay):
    print(f"Starting fetch: {url}")
    await asyncio.sleep(delay)      
    print(f"Finished fetch: {url}")
    return f"data from {url}"

async def main():
    results = await asyncio.gather(
        fetch_url("https://api.example.com/users", 2),
        fetch_url("https://api.example.com/posts", 5),
        fetch_url("https://api.example.com/comments", 3),
    )
    print(results)

await main()

Starting fetch: https://api.example.com/users
Starting fetch: https://api.example.com/posts
Starting fetch: https://api.example.com/comments
Finished fetch: https://api.example.com/users
Finished fetch: https://api.example.com/comments
Finished fetch: https://api.example.com/posts
['data from https://api.example.com/users', 'data from https://api.example.com/posts', 'data from https://api.example.com/comments']


---
# 6. Exceptions, `with` and JSON

Short section, but `with` matters more than it looks: it is the answer to
"where did my RAII go".

### 6.1 `try` / `except` / `else` / `finally`

In [ ]:
# ---------------- C++ : try / catch / throw (reference only) ----------------
#
#   try {
#       int x = std::stoi(input);              // throws std::invalid_argument
#       if (x == 0) throw std::runtime_error("zero");
#   }
#   catch (const std::invalid_argument& e) { ... }
#   catch (const std::runtime_error& e)    { ... }
#   catch (...)                            { ... }     // catch-all
#
#   // No else, no finally. Guaranteed cleanup comes from RAII destructors.
# ------------------------------------------------------------------------------

In [73]:
def divide(a, b):
    return a / b

try:
    x = int("42")                # ValueError if not a number
    result = divide(10, x)       # ZeroDivisionError if x == 0
except ValueError:
    print("bad input")
except ZeroDivisionError:
    print("division by zero")
except Exception as e:           # the catch-all; always name it and print it
    print(f"unexpected {type(e).__name__}: {e}")
else:
    print("no exception, result =", result)      # runs ONLY if nothing was raised
finally:
    print("finally: always runs")

# try changing this -> replace "42" with "0", then with "abc", and re-run

no exception, result = 0.23809523809523808
finally: always runs


> **Tip.** Python adds two clauses to the C++ model. `else` runs only when no exception was
> raised, which keeps the risky line and the follow-up logic clearly separated. `finally` always
> runs, even if the exception propagates.
>
> **Trap.** A bare `except:` catches everything, including `KeyboardInterrupt`, and hides real
> bugs. Write `except Exception as e:` and print `e`.

### 6.2 `with` keyword

In [4]:
from pathlib import Path

path = Path("notes.txt")

with open(path, "w") as f:          # 'w' truncates; closes at the end of the block
    f.write("line 1\n")
    f.write("line 2\n")

with open(path, "a") as f:          # 'a' appends
    f.write("line 3\n")

with open(path) as f:               # 'r' is the default
    for line in f:                  # iterating a file yields lines, memory-efficient
        print(" ", line.strip())    # strip() removes the trailing newline

print(path.exists(), path.suffix, path.stat().st_size, "bytes")
path.unlink()                       # delete
print("still there?", path.exists())

  line 1
  line 2
  line 3
True .txt 24 bytes
still there? False


> **Tip.** `with` is the direct replacement for scope-based cleanup. The file is closed when the
> block exits, including when an exception is thrown inside it. This is why you should never rely
> on `__del__`: object destruction is up to the garbage collector and is not guaranteed to happen
> when you expect.
>
> **Why this matters.** The same syntax appears well beyond files — database sessions, HTTP
> clients, and framework tracing callbacks all use `with`. Whenever a resource must be released,
> Python spells it `with`.

### 6.3 `json`: dicts in, text out

In [ ]:
import json

data = {"model": "demo", "params": {"temperature": 0.7}, "tags": ["a", "b"], "ok": True}

text = json.dumps(data)                   # dict -> string
print(text)
print(type(text))

back = json.loads(text)                   # string -> dict
print(back["params"]["temperature"], type(back))

print(json.dumps(data, indent=2))         # readable form, for logging

try:
    json.loads("{not valid}")
except json.JSONDecodeError as e:
    print("JSONDecodeError:", e.msg)

{"model": "demo", "params": {"temperature": 0.7}, "tags": ["a", "b"], "ok": true}
<class 'str'>
0.7 <class 'dict'>
{
  "model": "demo",
  "params": {
    "temperature": 0.7
  },
  "tags": [
    "a",
    "b"
  ],
  "ok": true
}
JSONDecodeError: Expecting property name enclosed in double quotes


> **Tip.** Note the mapping: JSON object to `dict`, array to `list`, `true` to `True`, `null` to
> `None`. This is why dicts are everywhere in this ecosystem — every API request and response is
> a JSON document, and `json.loads` hands you a dict directly, with no schema definition or
> parsing library involved.

---
### Check yourself: Section 6

**1.** When does the `else` block of a `try` statement run, and how is it different from putting
that code at the end of the `try` block?

<details><summary>Answer</summary>

It runs only when no exception was raised. The difference is scope of protection: code inside
`try` is covered by the handlers, so an error in your follow-up logic would be caught by the
wrong `except` clause. Code in `else` is not covered.
</details>

**2.** Why should you not use `__del__` to close a file?

<details><summary>Answer</summary>

Because Python gives no guarantee about when, or whether, `__del__` runs. There is no
deterministic destruction as in C++. Use `with`, which closes the resource at the end of the
block regardless of how the block exits.
</details>

**3.** What Python type does `json.loads('{"a": [1, 2], "b": null}')` return, and what is `"b"` bound to?

<details><summary>Answer</summary>

A `dict`, with `"a"` mapped to the list `[1, 2]` and `"b"` mapped to `None`.
</details>

### Question 6: JSON Logger
Write a program that:
1. Opens a file using `with`.
2. Reads a JSON configuration dictionary from it.
3. Updates a 'last_run' timestamp key.
4. Writes the updated dictionary back to the file in a readable (indented) format.

In [5]:
# write your code here
import json
from datetime import datetime

filename = "config.json"
try:
    with open(filename, "r") as f:
        config = json.load(f)        
except FileNotFoundError:
    config = {"app_name": "MyApp", "version": "1.0"}

config["last_run"] = datetime.now().isoformat()   

with open(filename, "w") as f:
    json.dump(config, f, indent=2)   

print(config)


{'app_name': 'MyApp', 'version': '1.0', 'last_run': '2026-09-14T09:56:40.210801'}


---
# Appendix A. C++ to Python Reference

| Concept | C++ | Python |
|---|---|---|
| Null value | `nullptr` | `None` |
| Booleans | `true` / `false` | `True` / `False` |
| Increment | `x++` | `x += 1` |
| Integer division | `x / y` on ints | `x // y` |
| Exponentiation | `pow(x, y)` | `x ** y` |
| String format | `std::format` / `printf` | `f"{var}"` |
| Logical operators | `&&` `\|\|` `!` | `and` `or` `not` |
| Block delimiter | `{ }` | indentation |
| Dynamic array | `std::vector` | `list` |
| Hash map | `std::unordered_map` | `dict` |
| Hash set | `std::unordered_set` | `set` |
| Ordered map | `std::map` | `dict` (insertion order, 3.7+) |
| Immutable sequence | `std::tuple` | `tuple` |
| Copy | copy constructor | `x.copy()` / `copy.deepcopy(x)` |
| Reference | `T&` | the default for every name |
| Constructor | `ClassName(...)` | `__init__` |
| Destructor | `~ClassName()` | none you should use |
| `this` | `this->member` | `self.member` |
| Virtual method | `virtual f()` | `def f()` (all are virtual) |
| Print form | `operator<<` | `__repr__` / `__str__` |
| Type check | `dynamic_cast` / `typeid` | `isinstance()` |
| Optional value | `std::optional<T>` | `Optional[T]` / `T \| None` |
| Struct of data | `struct S {}` | `TypedDict` / Pydantic / `@dataclass` |
| Enumerated values | `enum class` | `Literal[...]` / `Enum` |
| Try / catch | `try` / `catch` | `try` / `except` |
| Throw | `throw` | `raise` |
| Catch-all | `catch (...)` | `except Exception as e` |
| Guaranteed cleanup | RAII destructor | `with` / `finally` |
| Range for | `for (auto& x : v)` | `for x in v` |
| Index loop | `for (int i = 0; ...)` | `for i in range(n)` |
| Index and value | manual counter | `enumerate(v)` |
| Two sequences | manual index | `zip(a, b)` |
| Lambda | `[capture](p){ }` | `lambda p: expr` |
| Capture | explicit capture list | automatic |
| Default parameter | `void f(int x = 0)` | `def f(x=0):` |
| Variadic | `va_list` / templates | `*args` / `**kwargs` |
| Overloading | multiple signatures | not available |
| Header | `#include <x>` | `import x` |
| Namespace | `namespace ns {}` | module / package |
| Async | `std::async` / `std::future` | `async def` / `await` |
| Wait for many | `future.get()` in a loop | `asyncio.gather(...)` |
| Thread | `std::thread` | `asyncio` for I/O, `threading` for the rest |
| Smart pointer | `unique_ptr` / `shared_ptr` | garbage collected |
| Compile-time | `constexpr` / templates | nothing; it is interpreted |

---
# Appendix B. The Traps, Collected

Ranked roughly by how much time they cost a C++ programmer.

1. **`b = a` does not copy.** Two names, one object. Use `.copy()` or `copy.deepcopy()`.
   Mutating a list or dict argument inside a function changes the caller's object; there is no `const&`.
2. **Forgetting `await`.** You get a coroutine object rather than a result, and only a warning.
3. **Mutable default arguments.** `def f(x=[])` shares one list across all calls. Default to `None`.
4. **Type hints do nothing at runtime.** Only Pydantic actually validates.
5. **`d["missing"]` raises `KeyError`.** Use `d.get(key, default)`.
6. **Indentation is syntax.** Four spaces, never mixed with tabs.
7. **No `++` or `--`.** Use `x += 1`.
8. **`True` / `False` / `None` are capitalised**, and `None` is compared with `is`, not `==`.
9. **Strings are immutable.** Build new ones; assigning to `s[0]` is a `TypeError`.
10. **`{}` is an empty dict, not a set.** Use `set()`. And `(42)` is an int; `(42,)` is a tuple.
11. **No function overloading.** The second `def` of a name silently replaces the first.
12. **Assigning to a name inside a function makes it local.** Use `global` or `nonlocal` to reassign
    an outer variable, or you get `UnboundLocalError`.
13. **`int("3.14")` raises `ValueError`.** Use `int(float("3.14"))`.
14. **No destructors to rely on.** Use `with` for anything that must be released.
15. **In agent code, docstrings and `Field(description=...)` are sent to the model.** They are not
    comments; write them as instructions.

---
### You are done

Two things carry the most weight in what comes next: **types as runtime data** (Section 3) and
**async** (Section 5). If either felt shaky, that is the material to reread.